src/dom.ts

```diff
@@ -0,0 +1,68 @@
+const nanohtml = require('@tram-one/nanohtml');
+const rbel = require('@tram-one/rbel');
+const hyperx = require('@tram-one/hyperx');
+
+import { TRAM_HOOK_KEY } from './engine-names';
+import {
+	pushWorkingKeyBranch,
+	popWorkingKeyBranch,
+	incrementWorkingKeyBranch,
+	copyWorkingKey,
+	restoreWorkingKey,
+} from './working-key';
+import observeTag from './observe-tag';
+import processEffects from './process-effects';
+import { TRAM_TAG } from './node-names';
+
+import { Registry, Props, Children, DOMTaggedTemplateFunction } from './types';
+
+/**
+ * This function takes in a namespace and registry of custom components,
+ * and builds a `dom` template tag function that can take in a template XML string.
+ *
+ * This function shouldn't need to be called directly, instead, you can use `registerHtml` or `registerSvg`
+ *
+ * @param namespace namespace to create nodes in (by default XHTML namespace)
+ * @param registry mapping of tag names to component functions
+ */
+export const registerDom = (namespace: string, registry: Registry = {}): DOMTaggedTemplateFunction => {
+	// modify the registry so that each component function updates the hook working key
+	const hookedRegistry = Object.keys(registry).reduce((newRegistry, tagName) => {
+		const tagFunction = registry[tagName];
+		const hookedTagFunction = (props: Props, children: Children) => {
+			// push a new branch onto the working key so any values that need to be unique among components
+			// but consistent across renders can be read
+			const stringifiedProps = JSON.stringify(props);
+			const newBranch = `${tagName}[${stringifiedProps}]`;
+			pushWorkingKeyBranch(TRAM_HOOK_KEY, newBranch);
+
+			// increment branch so that we have a unique value (in case we are rendering a list of components)
+			incrementWorkingKeyBranch(TRAM_HOOK_KEY);
+			const uniqueBranch = copyWorkingKey(TRAM_HOOK_KEY);
+
+			// create a tag function that has the args passed in
+			const populatedTagFunction = () => {
+				// reset working key so we have the correct place when starting a new component
+				restoreWorkingKey(TRAM_HOOK_KEY, uniqueBranch);
+
+				return tagFunction(props, children);
+			};
```

Comment:
Previously this function used `(...args)` instead of `(props, children)` in method definitions. Potentially at the time I didn't know what this method definition would be, or wanted it to be more flexible, or maybe it was really _really_ early on when component methods had a ton of things in them. Either way, it looks like this now, and should be easier to understand and more explicit.